app/components/base_metadata_component.rb

```diff
@@ -0,0 +1,75 @@
+# frozen_string_literal: true
+
+class BaseMetadataComponent < ApplicationComponent
+  attr_reader :resource
+
+  def initialize(resource:)
+    @resource = decorate(resource)
+  end
+
+  private
+
+    def attributes_list
+      raise NotImplementedError
+    end
+
+    def decorate(resource)
+      raise NotImplementedError
+    end
+
+    def attributes
+      attributes_list
+        .map do |attr|
+          label = format_label(attr)
+          value = format_value(attr: attr, value: resource.send(attr))
+          [attr, label, value]
+        end
+        .reject { |_attr, _label, val| val.blank? }
+    end
+
+    def format_label(attr)
+      resource.to_model.class.human_attribute_name(attr)
+    end
+
+    def format_value(value:, attr:)
+      if value.is_a? Enumerable
+        format_multi_value(value: value, attr: attr)
+      elsif value.respond_to?(:strftime) # Date/Time/DateTime/TimeWithZone etc
+        value.to_formatted_s(:long)
+      elsif value.is_a? Authorship
+        value.display_name
+      elsif value.is_a? ApplicationComponent
+        render value
+      elsif attr == :related_url
+        format_link(value)
+      else
+        value.to_s
+      end
+    end
+
+    def format_multi_value(value:, attr:)
+      return nil if value.empty?
+
+      list_items =
+        value
+          .map { |member| format_value(value: member, attr: attr) }
+          .compact
+          .map { |member| %(<li class="multiple-member">#{member}</li>) }
+          .join
+
+      %(<ol class="multiple-values">#{list_items}</ol>).html_safe
```

Comment:
I'm not crazy about including the html in here, but I understand this is probably the simplest path.

My original idea for dealing with formatting individual fields differently, was to make separate components for each of them. See d953b965acf03c2cadf8e9c74283cd8c174719ec

So, you could make `DisplayMultiValueComponent` and `DisplayLinkComponent` classes for your `format_multi_value` and `format_link` methods respectively, and then the `format_value` method gets a little less branchy. Each case of a doi, multiple field, or link field, would be handled by the `elsif value.is_a? ApplicationComponent` case.

You'd also have to define the respective methods on the decorators to assign each field to its given component.

However, this does make things a bit more complicated... certainly more files. So, if we want to pass on this and try another direction in the future, I'm okay with that.

Reply:
The problem with doing that is, the items _within_ each multi-value field, could themselves be any number of datatypes which need to be formatted. The `format_multi_value` method recursively calls `format_value` on each of its items, so it needs to be able to call back into `BaseMetadataComponent.format_value`. Now, `format_value` is a pure function, so we could make that a class level method and allow other classes to call into it, but that seems like a major WTF in six months

Reply:
Ah, good point. Well, we may need to rethink this eventually. As different presentation options crop up, I don't want to end up putting all these different html formats in this component.